src/main/java/org/opentripplanner/graph_builder/impl/shapefile/NihShapefileStreetGraphBuilderImpl.java

```diff
@@ -0,0 +1,457 @@
+                    foundStart = true;
+                    startEdges.add(edge);
+                    farStartVertices.add(to);
+                } else if (toLabel.equals(startIntersection.getLabel())) {
+                    foundStart = true;
+                    startEdges.add(edge);
+                    farStartVertices.add(from);
+                }
+
+                if (fromLabel.equals(endIntersection.getLabel())) {
+                    foundEnd = true;
+                    endEdges.add(edge);
+                    farEndVertices.add(to);
+                } else if (toLabel.equals(endIntersection.getLabel())) {
+                    foundEnd = true;
+                    endEdges.add(edge);
+                    farEndVertices.add(from);
+                }
+            }
+
+            if (!foundStart || !foundEnd) {
+                LOG.warn("Could not find start or end for {}: {}, segment {}", wayLabel, wayName, segmentId);
+                continue;
+            }
+
+            // check if the start and end intersections are connected by a single graph segment
+            // (if so, there will be two edges found, one the reverse of the other)
+            boolean foundMatch = false;
+            for (StreetEdge start : startEdges) {
+                for (StreetEdge end : endEdges) {
+                    if (start.isEquivalentTo(end)) {
+                        foundMatch = true;
+                        matchedEdges.add(start);
+                    }
+                }
+            }
+
+            if (!foundMatch) {
+                // check if the start and end intersections are connected by two edges (that share a vertex in between)
+                // note the shared vertex, if found, in order to add edges that go to/from it later
+                String sharedVertex = null;
+                for (Vertex start : farStartVertices) {
+                    for (Vertex end : farEndVertices) {
+                        if (start.getLabel().equals(end.getLabel())) {
+                            sharedVertex = start.getLabel();
+                            break;
+                        }
+                    }
+                }
+
+                if (sharedVertex != null) {
+                    LOG.info("Start and end vertex share a midpoint at {}", sharedVertex);
+                    // go grab only edges that go between shared vertex and start or end edge
+                    List<StreetEdge> candidates = new ArrayList<>();
+                    candidates.addAll(startEdges);
+                    candidates.addAll(endEdges);
+                    for (StreetEdge edge: candidates) {
+                        if (edge.getFromVertex().getLabel().equals(sharedVertex) ||
+                                edge.getToVertex().getLabel().equals(sharedVertex)) {
+
+                            matchedEdges.add(edge);
+                        }
+                    }
+                } else {
```

Comment:
Is this extremely slow and/or memory-hungry? It seems like the code above is going to a lot of trouble to avoid doing this for the common cases, and I think it would make the code clearer to just default to this. If special-casing the 1-edge and 2-edge cases is necessary then it would be clearer to move some of the detection code out to its own function so you could do something like `if (one_edge()) { handle... } else if (two_edges()) { handle... } else { three edges...}`